jcl/src/java.base/share/classes/java/lang/Class.java

```diff
@@ -5070,25 +5073,58 @@ public boolean isRecord() {
 	public native boolean isSealed();
 
 	/**
-	 * Returns an array of permitted subclasses as ClassDesc objects related to the calling class.
+	 * Returns the permitted subclasses related to the calling sealed class as an array
+	 * of Class objects. If the calling class is not a sealed class, is an array class,
+	 * or is primitive, then this method returns null instead.
+	 * The order of any classes returned in the array is unspecified, and any classes
+	 * that cannot be loaded are not included in the returned array.
+	 * 
+	 * @return array of Class objects if permitted subclasses exist or null if not a sealed class.
 	 * 
-	 * @return array of ClassDesc objects.
-	 * For a class that is not sealed, an empty array is returned.
+	 * @throws SecurityException if access to any of the classes returned in the array is denied
 	 */
 	@CallerSensitive
-	public ClassDesc[] permittedSubclasses() {
+	public Class<?>[] getPermittedSubclasses() throws SecurityException {
 		if (!isSealed()) {
-			return new ClassDesc[0];
+			return null;
 		}
 
-		String[] permittedSubclassesNames = permittedSubclassesImpl();
-		ClassDesc[] permittedSubclasses = new ClassDesc[permittedSubclassesNames.length];
+		if (null == cachedPermittedSubclasses) {
+			String[] permittedSubclassesNames = permittedSubclassesImpl();
+			ArrayList<Class<?>> permittedSubclasses = new ArrayList<>(permittedSubclassesNames.length);
 
-		for (int i = 0; i < permittedSubclassesNames.length; i++) {
-			String subclassName = permittedSubclassesNames[i];
-			permittedSubclasses[i] = ClassDesc.of(permittedSubclassesNames[i]);
+			for (int i = 0; i < permittedSubclassesNames.length; i++) {
+				String subclassName = permittedSubclassesNames[i];
+				try {
+					Class<?> permitted = forNameImpl(subclassName, false, this.classLoader);
+					permittedSubclasses.add(permitted);
+				} catch (ClassNotFoundException e) {
+					// do nothing if class not found
+				}
+			}
+			cachedPermittedSubclasses = permittedSubclasses.toArray(new Class<?>[0]);
```

Comment:
This should use `getUnsafe().putObjectRelease()` to update the field. There are examples in this file.